### DISCUSSION TITLE
Is there a way to keep appending copied entries to the default clipboard (0)?

### DISCUSSION THREAD

**Discussion by xfzv** [Slackadays/Clipboard#110] [Q&A] [answered]
I read the README but I can't find any way to achieve this, maybe I'm missing it completely.

In KDE Plasma, the clipboard applet ([Klipper](https://userbase.kde.org/Klipper)) acts like this:

> The traditional expectation of a clipboard is that you can copy a section of text in order to paste it elsewhere. If you copy a section, the previous one is lost. Klipper solves this problem at a stroke. The item you copied last will still be the default one to be pasted, but others will have been stored in a buffer, so you can choose to paste your selections in a different order. Even the number of items stored in the buffer can be configured! Click on the icon on your tray or panel, and you will see the possible selections, as well as some options. Let's take a look at some configuration options. 

Is this possible to achieve something similar with Clipboard? basically: keep copying items to system clipboard with `Ctrl+c` or selection and then run `cb show` to list them all?

Currently, unless I'm mistaken, `cb show` only prints the last copied item, which overwrites the previous one.

**Comment by Slackadays** [comment, marked as answer]
Right now Clipboard has zero concept of history, which is something I'm thinking about how to implement. So, you won't be able to append anything in the meantime. However, once it's in, the history feature (probably) won't have the ludicrous limitation of many other clipboard managers where the history is limited to something like 1000 entries, so stay tuned! :)

**Reply by xfzv** [reply]
Good news! Will keep an eye on the releases.

Thank you.

**Reply by Slackadays** [reply]
@xfzv Basic history support is now available in the latest commit, would you mind trying it out? Right now you can make new entries by copying, cutting, or clearing, and you can see how many entries there are with the Info action and limit how many there can be with `CLIPBOARD_HISTORY`.

**Reply by xfzv** [reply]
Thanks for the heads up. I've just installed `clipboard-git` on the AUR to try it out.

```bash
❯ cb --help
│This is the Clipboard Project 0.7.1 (commit e7e0ef9), the cut, copy, and paste system for the command line.

❯ cb copy "Test"
✅ Copied text "Test"

❯ cb copy "Test2"
✅ Copied text "Test2"

❯  cb show
🔷 Here are the first 5 bytes in clipboard 0: Test2

❯ cb
┍━┫ All of your clipboards with content ┣━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┑
│ 0│ Test2                                                                           │
┕━━┷━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┙

❯  cb info
┍━┫ Info for clipboard 0 ┣━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┑                                                                                                                                                                                            
│ Stored in /tmp/Clipboard/0                                                                                                                                                                                                                                                                                                                                                                                                                             
│ Total entries: 2
```

The two entries are visible in `/tmp/Clipboard/0/data`.

However, I would have expected `cb` to display the two entries, not only the most recent one. Maybe it's just not implemented yet?

**Reply by Slackadays** [reply]
You're right, it doesn't show those two entries. However, although I would love to have that, I can't actually think of a good UI/UX that would make this possible without either using up a ton of space or not showing enough entries to be useful. If you have a potential solution, I'm all ears.

The latest commit now includes a basic History action (`cb history`) that will just show you everything (watch out if you have thousands of entries) and you can perform other actions on a specific entry by using the `--entry num` option where num is the entry number. The next step from here is going to be making History able to move an entry to be the latest.

By the way, thank you for testing this new feature! :smile: 

Edit: If you think about it, the Status, Show, and History actions are like one dimension each in a 3D clipboard storage system. Status shows you all your clipboards which is one dimension. Show shows you all items in a clipboard which is another dimension. Finally, History shows you all the entries in a clipboard, yet another dimension. Therefore, showing entries in the Status action with content would require some way of showing the data with a 3D aspect to it which is super tricky in a terminal and maybe possible in a GUI which is outside the scope of CB right now.

**Reply by xfzv** [reply]
I'm using [clipmenu](https://github.com/cdown/clipmenu) with [rofi](https://github.com/davatorium/rofi). `CM_LAUNCHER='rofi' CM_HISTLENGTH=10 clipmenu -p "Clipmenu "`:

![image](https://github.com/Slackadays/Clipboard/assets/78810647/540e0927-1f00-4b0e-8fbe-aeeb71b71bc8)

`CM_HISTLENGTH=10` limits the number of displayed entries but they're all available if scrolling with up/down arrow keys / mouse wheel or if searching. 

I tried piping `cb history` to rofi, [fzf](https://github.com/junegunn/fzf) and [bat](https://github.com/sharkdp/bat) but it doesn't work.

**Reply by Slackadays** [reply]
Currently the reason piping `cb history` doesn't work is because it detects that you're piping it, so it tries to do an action ~~that doesn't exist yet~~ (History JSON). There also isn't any good way to determine if you're using a JSON program or something else, so I don't know if you'll be able to use rofi, fzf, or bat without taking in a JSON input. Would some other setup or option setting work for you?

Edit: History JSON is now available

**Comment by Slackadays** [comment]
Just pinging @xfzv, @jstetten, @geomoetric, and @houman-babai because the release with unlimited history (0.8.0) is now out!

**Reply by jstetten** [reply]
Thanks, didn't know! This was the one feature I needed the most, so looking forward to try it out.

**Comment by geomoetric** [comment]
Hey @Slackadays I've been testing the new update with history and trying to figure out how to get `cb` to save everything I copy from the system clipboard from browser/non-terminal into my `cb 0` clipboard. I have the following settings enabled:
```
export CLIPBOARD_HISTORY=20
export CLIPBOARD_CUSTOMPERSIST=".*"
# export CLIPBOARD_CUSTOMPERSIST=0   ## Tried this one as well
# did not set CLIPBOARD_PERSISTDIR as the default ~/.clipboard/ works fine for my purposes
```
Please let me know if I'm missing something here. Thanks!

**Reply by Slackadays** [reply]
The `CLIPBOARD_CUSTOMPERSIST` is actually from the latest commits that will be made into a release tomorrow, so for now you'll have to set `CLIPBOARD_ALWAYS_PERSIST=1` in order to make it persistent. For saving everything from the system clipboard as soon as you copy it, check out issue #139. Otherwise, you'll need to run `cb` manually to get it to update the history.

**Reply by geomoetric** [reply]
Ok cool! I was using `CLIPBOARD_ALWAYS_PERSIST=1` previously, sounds like I'll use `export CLIPBOARD_CUSTOMPERSIST=".*"` for the same behavior in the next release.

I suppose my true issue isn't necessarily [#139](https://github.com/Slackadays/Clipboard/issues/139). `cb` always captures everything on the clipboard, but copying things too quickly won't record every persistent item. If I wait long enough between copying items to the system keyboard—say 10 seconds—then everything seems to capture and persist accurately in `cb`. If there's an adjustment or setting that can be made to enable faster capture from the system keyboard to the `cb` history I'd appreciate it—but obviously I have no idea how that would work!

**Reply by Slackadays** [reply]
> If I wait long enough between copying items to the system keyboard—say 10 seconds—then everything seems to capture and persist accurately in cb. If there's an adjustment that can be made to enable faster capture from the system keyboard to the cb history I'd appreciate it

So, the crux of the issue here is that `cb` doesn't actually update anything on its own in the background, but rather only when it runs. Therefore, waiting 10 seconds after copying something is actually ensuring that the system clipboard has all of the content ready to be served up, and `cb` queries that content when you run it. The latest update has a reprioritization of Wayland over X11, so if you're using Wayland, it might work better. If you'd like to try out the latest version and see if it's any better then you can just grab the latest commit from GHA or compile it.

**Reply by Slackadays** [reply]
@geomoetric just realized that you can actually get this right now if you install the flatpak. the flatpak has a special desktop entry that uses a script to show the clipboard status in the terminal and updates every few seconds. therefore, you get the traditional clipboard manager tracking functionality without having to do it yourself. https://flathub.org/apps/app.getclipboard.Clipboard

**Reply by geomoetric** [reply]
Cool! Currently on Mac so I won't be able to implement it as is, but seems like there could be an auto-update script that accomplishes the same thing. Eventually, maybe that's some sort of -i interactive mode that just keeps refreshing, but even a cron job might be able to do it. Thanks!!
